llvm/test/Analysis/CostModel/AArch64/sve-intrinsics.ll

```diff
@@ -930,8 +930,8 @@ define void @get_lane_mask() #0 {
 ; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of 16 for: %mask_v8i1_i32 = call <8 x i1> @llvm.get.active.lane.mask.v8i1.i32(i32 undef, i32 undef)
 ; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of 8 for: %mask_v4i1_i32 = call <4 x i1> @llvm.get.active.lane.mask.v4i1.i32(i32 undef, i32 undef)
 ; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of 4 for: %mask_v2i1_i32 = call <2 x i1> @llvm.get.active.lane.mask.v2i1.i32(i32 undef, i32 undef)
-; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of RThru:48 CodeSize:33 Lat:33 SizeLat:33 for: %mask_v32i1_i64 = call <32 x i1> @llvm.get.active.lane.mask.v32i1.i64(i64 undef, i64 undef)
-; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of RThru:6 CodeSize:5 Lat:5 SizeLat:5 for: %mask_v16i1_i16 = call <16 x i1> @llvm.get.active.lane.mask.v16i1.i16(i16 undef, i16 undef)
+; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of 64 for: %mask_v32i1_i64 = call <32 x i1> @llvm.get.active.lane.mask.v32i1.i64(i64 undef, i64 undef)
+; CHECK-VSCALE-1-NEXT:  Cost Model: Found costs of 32 for: %mask_v16i1_i16 = call <16 x i1> @llvm.get.active.lane.mask.v16i1.i16(i16 undef, i16 undef)
```

Comment:
This increase is worth looking into because the matching codegen tests look good. The new value looks like a default scalarisation cost? which we know shouldn't apply.

Reply:
This is coming from `AArch64TTIImpl::getIntrinsicInstrCost`, which returns a high cost for fixed-vector types which are not legal even if `shouldExpandGetActiveLaneMask` returns false:

```
if (!getTLI()->shouldExpandGetActiveLaneMask(RetVT, OpVT) &&
    !getTLI()->isTypeLegal(RetVT)) {
    // We don't have enough context at this point to determine if the mask
    // is going to be kept live after the block, which will force the vXi1
    // type to be expanded to legal vectors of integers, e.g. v4i1->v4i32.
    // For now, we just assume the vectorizer created this intrinsic and
    // the result will be the input for a PHI. In this case the cost will
    // be extremely high for fixed-width vectors.
```

Reply:
Yeah I had to add this code a while back in order to prevent terrible code quality when tail-folding for fixed-width vectors. It really depends upon how the result is used as to what codegen you get, and given the primary generator of this intrinsic is the loop vectoriser it while be used as the incoming value for a PHI node.

Reply:
It does seem odd that the cost has gone up though.

Reply:
We should fix it though as part of this PR? I mean, there's no good reason to jump from 6 to 32 given the cost is just "SVE cost plus a sign extend".

Reply:
Oh sorry it makes perfect sense, because prior to this patch `shouldExpandGetActiveLaneMask` returned true for result type <32 x i1> or input operand i16, whereas now it returns false. So the code path has indeed changed. The increase in cost means one of two things:

1. The BasicTTIImpl calculation was too optimistic, or
2. The new cost is too high.

It might be interesting to hand-write a vectorised tail-folded loop using both of these intrinsic variants to see what happens?

Reply:
As discussed offline, for now I've changed `shouldExpandGetActiveLaneMask()` to only return false for fixed-length types if the result & operand types are already legal. This means that the costs which increased in `sve-intrinsics.ll` are now unchanged.